### PR TITLE
Stackless coroutines now can refer to themselves (through get_self() and friends)

### LINKED ISSUE #4155: get_self_id() for stackless threads returns invalid_thread_id
## Expected Behavior

`threads::get_self_id()` for stackless threads should return a valid value.

### PR DESCRIPTION
**Pull request by hkaiser**
Fixes #4155

### CODE REVIEW

**Comment by hkaiser** [comment]
@tianyi93 this should resolve your issues with stackless threads you were seeing in hpxMP. Please verify.

**Comment by tianyi93** [comment]
@hkaiser  It seems  ```error: 'hpx/coroutines/detail/coroutine_stackless_self.hpp' file not found``` is missing

**Comment by tianyi93** [comment]
@hkaiser SIGSEGV is happening when executing create_thread function under hpx/runtime/threads/detail/create_thread.hpp

File: libs/coroutines/include/hpx/coroutines/detail/coroutine_self.hpp [hunk 1]
```diff
@@ -129,67 +117,33 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
             return tmp;
         }
 
-        thread_id_type get_thread_id() const
-        {
-            HPX_ASSERT(m_pimpl);
-            return m_pimpl->get_thread_id();
-        }
+        virtual ~coroutine_self() = default;
```
Comment (msimberg) [inline libs/coroutines/include/hpx/coroutines/detail/coroutine_self.hpp]: Do you know if making all these functions virtual adds any noticeable overhead?

Comment (hkaiser) [reply libs/coroutines/include/hpx/coroutines/detail/coroutine_self.hpp]: I have not seen any significant slowdown from those, but I have not done any exhaustive measurements,

**Comment by msimberg** [comment]
@tianyi93 I guess your problem has already been fixed by @hkaiser? Not sure if the problem showed up on our builders in the first place but they are happy (enough) now.

**Comment by tianyi93** [comment]
@msimberg Yes, the compiler error has been fixed. The issue remains is the implementation in the hpxMP side now. Thanks!

**Pull request merged**